pkg/code-generator/codegen/templates/snapshot_template.go

```diff
@@ -80,6 +93,103 @@ func (s {{ .GoName }}Snapshot) HashFields() []zap.Field {
 	return append(fields, zap.Uint64("snapshotHash",  snapshotHash))
 }
 
+{{ if ge $num_of_hashable_resources 1}}
+
+func (s *{{ .GoName }}Snapshot) GetResourcesList(resource resources.Resource) (resources.ResourceList, error) {
+	switch resource.(type) {
+{{- range .Resources }}
+{{- if not .IsCustom }}
+	case *{{ .ImportPrefix }}{{ .Name }}:
+		return s.{{ upper_camel .PluralName }}.AsResources(), nil
+{{- end }}
+{{- end }}
+	default:
+		return resources.ResourceList{}, eris.New("did not contain the input resource type returning empty list")
+	}
+}
+
+func (s *{{ .GoName }}Snapshot) AddToResourceList(resource resources.Resource) error {
+	switch typed := resource.(type) {
+{{- range .Resources }}
+{{- if not .IsCustom }}
+	case *{{ .ImportPrefix }}{{ .Name }}:
+		s.{{ upper_camel .PluralName }} = append(s.{{ upper_camel .PluralName }}, typed)
+		s.{{ upper_camel .PluralName }}.Sort()
+		return nil
+{{- end }}
+{{- end }}
+	default:
+		return eris.New("did not add the input resource type because it does not exist")
+	}
+}
+
+func (s *{{.GoName}}Snapshot) ReplaceResource(i int, resource resources.Resource) error {
```

Comment:
I'm a little confused by this API. We are exposing the ability to replace a resource in a snapshot at a particualr index. How does the consumer of this library determine what i should be? 

Can you help me understand the use case where this would be important? I fear that our snapshots originally had a contract that they wouldn't be modified, so I want to make sure that if we're introducing support for that we consider all the eeffects

Reply:
So in gloo we make a clone of the snapshot, and do the same thing. We will replace the resource in the list or add to the list based off if the resource we are validating already exists in the snapshot.
This is done in the [apply function](https://github.com/solo-io/gloo/blob/master/projects/gateway/pkg/validation/validator.go#L470)